File: extractors/typescript/src/options.ts
import { basename } from "node:path";
import type { ProgressMode } from "./progress.js";
import type { Repository } from "./model/model.js";

/**
 * The extractor command-line contract (schemas/README.md §8), shared with the
 * Java jar and the C# binary so `codegraph snapshots --extractor` can drive
 * any of them, plus the three TypeScript flags of PLAN.md §14.5.
 */
export interface Options {
  readonly sources: readonly string[];
  readonly out: string;
  readonly progress: ProgressMode;
  readonly repository: Repository | undefined;
  /** A tsconfig path, `"none"` for the synthesized defaults, `undefined` to search. */
  readonly tsconfig: string | undefined;
  readonly allowJs: boolean;
  readonly ignoreNodeModules: boolean;
  readonly help: boolean;
  readonly version: boolean;
}

export class UsageError extends Error {
  override name = "UsageError";
}

const HTTPS_REMOTE = /^https:\/\/(?![^\s]*\.git$)[^\s?#]*[^\s?#/]$/;
const SHA = /^[0-9a-f]{7,64}$/;
const RELATIVE_PATH = /^$|^(?!\.\.?(?:\/|$))[^/\s]+(?:\/(?!\.\.?(?:\/|$))[^/\s]+)*$/;

export function parseOptions(args: readonly string[], cwd: string): Options {
  const sources: string[] = [];
  let out: string | undefined;
  let progress: ProgressMode = "auto";
  let remote: string | undefined;
  let commit: string | undefined;
  let root: string | undefined;
  let provider: string | undefined;
  let tsconfig: string | undefined;
  let allowJs = false;
  let ignoreNodeModules = false;
  let help = false;
  let version = false;

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i] as string;
    const value = (): string => {
      const next = args[i + 1];
      if (next === undefined) throw new UsageError(`${arg} needs a value`);
      i += 1;
      return next;
    };
    switch (arg) {
      case "--src":
        sources.push(value());
        break;
      case "--out":
        out = value();
        break;
      case "--progress": {
        const mode = value();
        if (mode !== "auto" && mode !== "plain" && mode !== "none") {
          throw new UsageError(`--progress must be auto, plain or none, got: ${mode}`);
        }
        progress = mode;
        break;
      }
      case "--no-progress":
        progress = "none";
        break;
      case "--repo-remote":
        remote = value();
        break;
      case "--repo-commit":
        commit = value();
        break;
      case "--repo-root":
        root = value();
        break;
      case "--repo-provider":
        provider = value();
        break;
      case "--tsconfig":
        tsconfig = value();
        break;
      case "--allow-js":
        allowJs = true;
        break;
      case "--ignore-node-modules":
        ignoreNodeModules = true;
        break;
      case "--help":
      case "-h":
        help = true;
        break;
      case "--version":
        version = true;
        break;
      default:
        throw new UsageError(`unknown option: ${arg}`);
    }
  }

  let repository: Repository | undefined;
  if (remote !== undefined || commit !== undefined || root !== undefined || provider !== undefined) {
    if (remote === undefined || commit === undefined || root === undefined) {
      throw new UsageError(
        "--repo-remote, --repo-commit and --repo-root go together (--repo-provider is optional)",
      );
    }
    if (!HTTPS_REMOTE.test(remote)) {
      throw new UsageError(`--repo-remote must be a normalized https URL (no ssh form, no .git): ${remote}`);
    }
    if (!SHA.test(commit)) throw new UsageError(`--repo-commit must be a lowercase hex sha: ${commit}`);
    if (!RELATIVE_PATH.test(root)) {
      throw new UsageError(`--repo-root must be a path relative to the repository root: ${root}`);
    }
    if (provider !== undefined && provider !== "github" && provider !== "gitlab") {
      throw new UsageError(`--repo-provider must be github or gitlab, got: ${provider}`);
    }
    repository = {
      remote,
      commit,
      root,
      ...(provider === undefined ? {} : { provider }),
    };
  }

  if (sources.length === 0) sources.push(".");
  return {
    sources,
    out: out ?? `${basename(cwd)}-codegraph.jsonl`,
    progress,
    repository,
    tsconfig,
    allowJs,
    ignoreNodeModules,
    help,
    version,
  };
}

export function usage(version: string): string {
  return `codegraph-typescript ${version} — TypeScript extractor on the compiler API

USAGE
  codegraph-typescript [--src <dir>…] [--out <file>]

OPTIONS
  --src <dir>           source root to analyze; repeatable. Default: the current
                        directory. With several roots, anchors are relative to
                        their deepest common ancestor, which becomes the model's root.
  --out <file>          where to write the model. Default: <current-dir>-codegraph.jsonl
  --progress <mode>     auto (default: one line per phase on a terminal, nothing
                        when piped), plain, or none
  --no-progress         same as --progress none
  --tsconfig <file>     the tsconfig.json whose RESOLUTION options apply (paths,
                        baseUrl, rootDirs, jsx, lib, target, allowJs); default: the
                        nearest tsconfig.json at or above each root, the first root's
                        winning on conflict; \`none\` for the synthesized defaults
  --allow-js            also walk *.js/*.jsx/*.mjs/*.cjs (JSDoc types feed declaredType)
  --ignore-node-modules never read node_modules, even when present: every external
                        package is an unresolved stub module
  --repo-remote <url>   normalized https clone URL, no .git suffix
  --repo-commit <sha>   the sha this tree is at — a permalink, not a branch
  --repo-root <path>    repo-relative path of the analyzed root ("" at the repo root)
  --repo-provider <p>   github or gitlab, only when the hostname does not say
  --version             print the extractor version
  --help                this text

Nothing is built: every *.ts/*.tsx/*.mts/*.cts under the roots (node_modules, .git
and dist excluded) goes into ONE program; a tsconfig is read for resolution options
only; a package that is not installed is a stub. Exit codes: 0 ok, 1 failure,
2 usage, 3 unimplemented.
`;
}

File: extractors/elixir/lib/codegraph_elixir/options.ex
defmodule CodegraphElixir.Options do
  @moduledoc """
  The extractor command-line contract (schemas/README.md §8), shared with the
  Java jar, the C# binary and the TypeScript bin so `codegraph snapshots
  --extractor` can drive any of them, plus the Elixir flags of PLAN.md §16.5.
  """

  defstruct sources: [],
            out: nil,
            progress: :auto,
            repository: nil,
            deps: nil,
            trace: nil,
            explain_dropped: false,
            help: false,
            version: false

  @type t :: %__MODULE__{}

  @https_remote ~r"^https://(?![^\s]*\.git$)[^\s?#]*[^\s?#/]$"
  @sha ~r"^[0-9a-f]{7,64}$"
  @relative_path ~r"^$|^(?!\.\.?(?:/|$))[^/\s]+(?:/(?!\.\.?(?:/|$))[^/\s]+)*$"

  @spec parse([String.t()], String.t()) :: {:ok, t()} | {:error, String.t()}
  def parse(argv, cwd) do
    with {:ok, acc} <-
           walk(argv, %{
             sources: [],
             out: nil,
             progress: :auto,
             repo: %{},
             deps: nil,
             trace: nil,
             explain_dropped: false,
             help: false,
             version: false
           }),
         {:ok, repository} <- repository(acc.repo) do
      sources = if acc.sources == [], do: ["."], else: Enum.reverse(acc.sources)

      {:ok,
       %__MODULE__{
         sources: sources,
         out: acc.out || "#{Path.basename(cwd)}-codegraph.jsonl",
         progress: acc.progress,
         repository: repository,
         deps: acc.deps,
         trace: acc.trace,
         explain_dropped: acc.explain_dropped,
         help: acc.help,
         version: acc.version
       }}
    end
  end

  defp walk([], acc), do: {:ok, acc}
  defp walk(["--src", value | rest], acc), do: walk(rest, %{acc | sources: [value | acc.sources]})
  defp walk(["--out", value | rest], acc), do: walk(rest, %{acc | out: value})

  defp walk(["--progress", value | rest], acc) do
    case value do
      "auto" -> walk(rest, %{acc | progress: :auto})
      "plain" -> walk(rest, %{acc | progress: :plain})
      "none" -> walk(rest, %{acc | progress: :none})
      other -> {:error, "--progress must be auto, plain or none, got: #{other}"}
    end
  end

  defp walk(["--no-progress" | rest], acc), do: walk(rest, %{acc | progress: :none})
  defp walk(["--repo-remote", value | rest], acc), do: walk(rest, put_in(acc.repo[:remote], value))
  defp walk(["--repo-commit", value | rest], acc), do: walk(rest, put_in(acc.repo[:commit], value))
  defp walk(["--repo-root", value | rest], acc), do: walk(rest, put_in(acc.repo[:root], value))
  defp walk(["--repo-provider", value | rest], acc), do: walk(rest, put_in(acc.repo[:provider], value))
  defp walk(["--deps", value | rest], acc), do: walk(rest, %{acc | deps: value})
  defp walk(["--trace", value | rest], acc), do: walk(rest, %{acc | trace: value})
  defp walk(["--explain-dropped" | rest], acc), do: walk(rest, %{acc | explain_dropped: true})
  defp walk([flag | rest], acc) when flag in ["--help", "-h"], do: walk(rest, %{acc | help: true})
  defp walk(["--version" | rest], acc), do: walk(rest, %{acc | version: true})

  defp walk([flag], _acc)
       when flag in ~w(--src --out --progress --repo-remote --repo-commit --repo-root --repo-provider --deps --trace),
       do: {:error, "#{flag} needs a value"}

  defp walk([other | _], _acc), do: {:error, "unknown option: #{other}"}

  # `--repo-remote`, `--repo-commit` and `--repo-root` go together; copied
  # verbatim into the header after a shape check — the extractor invents nothing.
  defp repository(repo) when map_size(repo) == 0, do: {:ok, nil}

  defp repository(repo) do
    cond do
      not (Map.has_key?(repo, :remote) and Map.has_key?(repo, :commit) and Map.has_key?(repo, :root)) ->
        {:error, "--repo-remote, --repo-commit and --repo-root go together (--repo-provider is optional)"}

      not Regex.match?(@https_remote, repo.remote) ->
        {:error, "--repo-remote must be a normalized https URL (no ssh form, no .git): #{repo.remote}"}

      not Regex.match?(@sha, repo.commit) ->
        {:error, "--repo-commit must be a lowercase hex sha: #{repo.commit}"}

      not Regex.match?(@relative_path, repo.root) ->
        {:error, "--repo-root must be a path relative to the repository root: #{repo.root}"}

      Map.has_key?(repo, :provider) and repo.provider not in ["github", "gitlab"] ->
        {:error, "--repo-provider must be github or gitlab, got: #{repo.provider}"}

      true ->
        base = [{"remote", repo.remote}, {"commit", repo.commit}, {"root", repo.root}]
        {:ok, if(Map.has_key?(repo, :provider), do: base ++ [{"provider", repo.provider}], else: base)}
    end
  end

  def usage(version) do
    """
    codegraph-elixir #{version} — Elixir extractor on the compiler's parser

    USAGE
      codegraph-elixir [--src <dir>…] [--out <file>]

    OPTIONS
      --src <dir>           source root to analyze; repeatable. Default: the current
                            directory. With several roots, anchors are relative to
                            their deepest common ancestor, which becomes the model's root.
      --out <file>          where to write the model. Default: <current-dir>-codegraph.jsonl
      --progress <mode>     auto (default: one line per phase on a terminal, nothing
                            when piped), plain, or none
      --no-progress         same as --progress none
      --deps <dir>          parse dependency sources under <dir> for their EXPORTS only
                            (names and arities); keys never change, only counts do.
                            Default: none, even when deps/ exists beside the roots
      --trace <file>        merge the compiler trace `mix codegraph.trace` wrote inside the
                            project: calls the parser cannot see (macro-injected) become
                            `generated` edges; keys never change
      --explain-dropped     list every dropped site on stderr (reason, file:line, name/arity)
      --repo-remote <url>   normalized https clone URL, no .git suffix
      --repo-commit <sha>   the sha this tree is at — a permalink, not a branch
      --repo-root <path>    repo-relative path of the analyzed root ("" at the repo root)
      --repo-provider <p>   github or gitlab, only when the hostname does not say
      --version             print the extractor version
      --help                this text

    Nothing is compiled: every *.ex/*.exs under the roots (_build, deps, .git and
    node_modules excluded) is parsed with the compiler's own parser; names resolve
    through a lexical scope and an embedded table of what the BEAM ships; a module
    the corpus does not declare is a stub. Exit codes: 0 ok, 1 failure, 2 usage,
    3 unimplemented.
    """
  end
end
